### PR TITLE
feat(connection-form): support for multiple kms options from same provider COMPASS-8082

### PR DESCRIPTION
**Pull request by mabaasit**
<details>
<summary>Preview</summary>

image

</details>

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-connections/src/stores/connections-store-redux.ts
Comment (addaleax) [inline packages/compass-connections/src/stores/connections-store-redux.ts:1]: Note quite sure where to put this, so just leaving it here on the first file:

Screenshot 2024-08-29 at 19 01 49

(For a connection configured with 2 `local` providers, `local` and `local:1`)

Clicking the empty button crashes the app

File: packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+function getNextKmsProviderName<T extends KMSProviderType>(
+  kmsProviderType: T,
+  currentProviders: string[]
+): KMSProviderName<T> {
+  const currentNums = currentProviders
+    .map((name) => parseInt(name.split(':')[1], 10))
+    .filter((num) => !isNaN(num));
+  const name = currentNums.length === 0 ? 1 : Math.max(...currentNums) + 1;
+  return `${kmsProviderType}:${name}`;
```
Comment (addaleax) [inline packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx:44]: I don't know if we can select the name for the user here. The name _is_ persisted in the key vault collection, and while I'm not entirely sure, I would guess that it has to match between different usages of the same key (i.e. the name used by the application `kmsProviders` config should match the one used here).

I wanted to verify this myself but ran into some issues with my local FLE setup that I'll probably need to look into first

**Review by addaleax — COMMENTED** [review]
Left two comments before diving into a bigger review, hopefully tomorrow 🙂

File: packages/compass-connections/src/stores/connections-store-redux.ts
Comment (mabaasit) [reply packages/compass-connections/src/stores/connections-store-redux.ts:1]: actually i did not know that user can choose kms provider here. thanks for catching this.

File: packages/databases-collections/src/modules/create-namespace.ts [hunk 2]
```diff
@@ -168,13 +168,18 @@ const reducer: Reducer<CreateNamespaceState, Action> = (
...
     return { ...INITIAL_STATE };
   }
 
+  if (isAction<CloseAction>(action, CreateNamespaceActionTypes.Close)) {
+    // When a modal is closed, we should not clear the connectionMetaData
+    return {
+      ...INITIAL_STATE,
+      connectionMetaData: state.connectionMetaData,
+    };
+  }
```
Comment (mabaasit) [inline packages/databases-collections/src/modules/create-namespace.ts:181]: Drive-by: when opening modal to create a collection after closing it first time, the UI did not show encryption options, because we set `connectionMetaData` to initial state (`{}`).

File: packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx (hunk 1 above)
Comment (mabaasit) [reply packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx]: I updated the PR to add support for renaming kms. Currently it allows users to edit the name while warning them about:
1. Empty name
2. Duplicate name

In these cases, if a user proceeds to connect, currently it considers the last valid name (as we do not rename). 

It hides the full name of the provider (the prefix part) - e.g if the user types local123, the actual name being `local:local123`, we hide the `local:`. I am waiting design input on how to communicate this to the user.

File: packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  currentProviders: string[] = []
+): KMSProviderName<T> {
+  // For name, we are prefixing it with the type of the provider separated by a colon.
+  // This is because the name of the kms provider should always have the type of the provider
+  // and we are not showing it to the user and it is used internally.
+  if (currentProviders.length === 0) {
+    return `${kmsProviderType}:${kmsProviderType}1`;
+  }
+  const currentNums = currentProviders // local:local1
+    .map((name) => name.split(':')[1]?.replace(kmsProviderType, '')) // '1'
```
Comment (addaleax) [inline packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx:33]: The `[1]` here works because KMS provider names need to follow a `^<type>:[a-zA-Z0-9_]+$` scheme – should we validate that the name part only consists of `[a-zA-Z0-9_]+` to ensure that it doesn't contain a `:`?

File: packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    .map((x) => parseInt(x, 10)) // 1
+    .filter((x) => !isNaN(x));
+  const nextNum = currentNums.length === 0 ? 1 : Math.max(...currentNums) + 1;
```
Comment (addaleax) [inline packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx:36]:
```suggestion
  const nextNum = Math.max(0, ...currentNums) + 1;
```

File: packages/connection-form/src/utils/csfle-handler.ts [hunk 5]
```diff
@@ -152,29 +178,25 @@ export function handleUpdateCsfleKmsTlsParam({
...
   } else {
     tls[action.key] = action.value;
   }
   const tlsOptions = autoEncryption.tlsOptions ?? {};
-  if (Object.keys(tls).length === 0) {
-    delete tlsOptions[action.kmsProvider];
-  } else {
-    tlsOptions[action.kmsProvider] = tls;
-  }
+  tlsOptions[action.kmsProviderName] = tls;
```
Comment (addaleax) [inline packages/connection-form/src/utils/csfle-handler.ts:189]: Our of curiousity, what's bringing on this change? If the KMS provider name changes, does this "leak" TLS options that do not correspond to any of the KMS providers?

File: packages/connection-info/src/connection-secrets.ts [hunk 6]
```diff
@@ -190,10 +181,75 @@ export function extractSecrets(connectionInfo: Readonly<ConnectionInfo>): {
...
 
-function omitPropertiesWhoseValuesAreEmptyObjects<
-  T extends Record<string, Record<string, unknown>>
->(obj: T): Partial<T> {
+function omitPropertiesWhoseValuesAreEmptyObjects<T extends Document>(obj: T) {
   return Object.fromEntries(
     Object.entries(obj).filter(([, value]) => Object.keys(value).length > 0)
-  ) as Partial<T>;
+  );
+}
```
Comment (addaleax) [inline packages/connection-info/src/connection-secrets.ts:188]: This just removes the return type annotation and reduces type safety overall – any particular reason for that?

File: packages/connection-info/src/connection-secrets.ts [hunk 7]
```diff
@@ -190,10 +181,75 @@ export function extractSecrets(connectionInfo: Readonly<ConnectionInfo>): {
...
+
+function extractAutoEncryptionSecrets(data: Document) {
```
Comment (addaleax) [inline packages/connection-info/src/connection-secrets.ts:190]:
```suggestion
function extractAutoEncryptionSecrets(data: AutoEncryptionOptions) {
```

and this should have a return type annotation as well

File: packages/connection-info/src/connection-secrets.ts [hunk 8]
```diff
@@ -190,10 +181,75 @@ export function extractSecrets(connectionInfo: Readonly<ConnectionInfo>): {
...
+  function flattenObject(
+    obj: Document,
+    path = '',
+    result: Document = {}
+  ): Record<string, unknown> {
+    _.forEach(obj, (value, key) => {
+      key = key.replaceAll('.', '\\.');
```
Comment (addaleax) [inline packages/connection-info/src/connection-secrets.ts:197]: This probably works out okay because our keys do not contain literal `\.` sequences, but this code overall isn't the easiest to read – I don't think there's anything requiring us to stick to a paradigm where we specify the set of paths that we want to redact as a list of strings, manually redacting these objects the way we do for other sub-properties of `ConnectionOptions` is probably just fine too and might end up being both more readable and more type-safe

File: packages/connection-form/src/utils/csfle-handler.ts (hunk 5 above)
Comment (mabaasit) [reply packages/connection-form/src/utils/csfle-handler.ts:189]: ah, this i missed completely when renaming. i'll ensure in rename, tlsoption is also renamed. 
and this should remain as original code. will revert this change.

File: packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-card.tsx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+  const [hoverProps, isHovered] = useHoverState();
+  const [cardName, setCardName] = useState(kmsProviderName);
+  const [nameError, setNameError] = useState<string | undefined>();
+
+  const onRenameKmsProvider = useCallback(
+    (name: string) => {
+      const newName = `${kmsProviderType}:${name}` as KMSProviderName<T>;
+      setCardName(newName);
+      if (name === '') {
+        setNameError('Name cannot be empty');
```
[the remark this answers is not in the document]
Comment (mabaasit) [reply packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-card.tsx:70]: this was enforcing all the kms provider names to be prefixed by the type. i'll remove this check and allow users to have kms type as name.

File: packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx (hunk 3 above)
Comment (mabaasit) [reply packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx:33]: good point, should we actually force this pattern (`^<type>:[a-zA-Z0-9_]+$`) on the name of the kms? it will also simply extracting secrets?

Comment (addaleax) [reply packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-content.tsx:33]: I think it's okay to do that at least through TextInput validation if we can, yeah

File: packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-card.tsx [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+        connectionOptions={connectionOptions}
+        updateConnectionFormField={updateConnectionFormField}
+        kmsProviderType={kmsProviderType}
+        kmsProviderName={kmsProviderName}
+        fields={fields}
+        clientCertIsOptional={clientCertIsOptional}
+        noTLS={noTLS}
+      />
+    </Card>
+  );
```
Comment (addaleax) [inline packages/connection-form/src/components/advanced-options-tabs/csfle-tab/kms-provider-card.tsx:241]: Just wanted to say, I tried this out locally and it all feels very smooth and intuitive! This is amazing work ✨

File: packages/connection-form/src/utils/csfle-handler.ts [hunk 11]
```diff
@@ -184,10 +206,55 @@ export function handleUpdateCsfleKmsTlsParam({
...
+    };
+  }
+
+  function filterEmptyValues(obj: Record<string, any> | undefined) {
+    const values = Object.fromEntries(
+      Object.entries(obj ?? {})
+        .filter(([, v]) => Object.keys(v ?? {}).length > 0)
+        .map(([k, v]) => [k, v])
+    );
+    return Object.keys(values).length > 0 ? values : undefined;
```
Comment (addaleax) [inline packages/connection-form/src/utils/csfle-handler.ts:232]: Nit, we can use generics here to improve type safety:

```suggestion
  function filterEmptyValues<T extends object>(obj: T | undefined): Partial<T> | undefined {
    const values = Object.fromEntries(
      Object.entries(obj ?? {})
        .filter(([, v]) => Object.keys(v ?? {}).length > 0)
    );
    return Object.keys(values).length > 0 ? values as Partial<T> : undefined;
```

File: packages/connection-info/src/connection-secrets.ts [hunk 12]
```diff
@@ -190,10 +181,62 @@ export function extractSecrets(connectionInfo: Readonly<ConnectionInfo>): {
...
+function extractAutoEncryptionSecrets(data: AutoEncryptionOptions): {
+  data: AutoEncryptionOptions & AutoEncryptionKMSAndTLSOptions;
+  secrets: AutoEncryptionKMSAndTLSOptions;
+} {
+  const secrets: AutoEncryptionKMSAndTLSOptions = {};
+  // Secrets are stored in a kmsProviders and tlsOptions
+  const { kmsProviders, tlsOptions, ...result } = data;
+
+  for (const kmsProviderName in kmsProviders ?? {}) {
+    const key = kmsProviderName.match(/\b(local|aws|gcp|azure)\b(?:)?/)?.[1] as
```
Comment (addaleax) [inline packages/connection-info/src/connection-secrets.ts:210]: Just a question – Does

```suggestion
    const key = kmsProviderName.split(':')[0] as
```

work here?

Currently, this would end up matching something like `kmip:local` as `local`, which hopefully won't be a common case, but maybe this can just be avoided this way?

File: packages/connection-info/src/connection-secrets.ts [hunk 13]
```diff
@@ -190,10 +181,62 @@ export function extractSecrets(connectionInfo: Readonly<ConnectionInfo>): {
...
+
+function extractAutoEncryptionSecrets(data: AutoEncryptionOptions): {
+  data: AutoEncryptionOptions & AutoEncryptionKMSAndTLSOptions;
+  secrets: AutoEncryptionKMSAndTLSOptions;
+} {
+  const secrets: AutoEncryptionKMSAndTLSOptions = {};
+  // Secrets are stored in a kmsProviders and tlsOptions
+  const { kmsProviders, tlsOptions, ...result } = data;
+
+  for (const kmsProviderName in kmsProviders ?? {}) {
```
Comment (addaleax) [inline packages/connection-info/src/connection-secrets.ts:209]:
```suggestion
  for (const kmsProviderName of Object.keys(kmsProviders ?? {})) {
```

Not a huge deal, but `for ... in` is generally considered a bit less reliable because it walks the prototype chain, so while that doesn't make a big difference here, it's something we shouldn't start making common use of in the codebase

**Review by addaleax — COMMENTED** [review]
🚀

File: packages/connection-info/src/connection-secrets.ts (hunk 12 above)
Comment (mabaasit) [reply packages/connection-info/src/connection-secrets.ts]: fixed in [3f954e6](https://github.com/mongodb-js/compass/pull/6166/commits/3f954e6b6a3fb85f4943a929b5c551ff8901bb29)

**Review by addaleax — APPROVED** [review]
Looks good, feel free to consider the remaining unresolved comments optional 👍

File: packages/connection-form/src/utils/csfle-handler.ts (hunk 5 above)
Comment (mabaasit) [reply packages/connection-form/src/utils/csfle-handler.ts]: reverted in 1b179f689066e3d4948304a2345e0fad9335c7fa

File: packages/connection-info/src/connection-secrets.ts (hunk 6 above)
Comment (addaleax) [reply packages/connection-info/src/connection-secrets.ts:188]: I mean the return type annotation is still missing but that's not a big deal

Comment (mabaasit) [reply packages/connection-info/src/connection-secrets.ts:188]: 🤦🏽
